Clippy fixes + Clippy in CI

bschwind opened the pull request:
This was essentially just a run of `cargo clippy --fix --release --all-targets`, plus an addition to CI to run these clippy checks for pull requests and commits.

I also moved the constructor functions for `Primitive`, `Array`, and `Object` into value as `Value::new_primitive()`, `Value::new_array()`, and `Value::new_object()`, since those functions all returned `Value`.

There should be no functional library changes introduced in this PR, verified with `cargo test`.

bschwind committed:
Apply clippy fixes

bschwind committed:
Run clippy with errors on warnings in CI

bschwind commented:
I hopefully signed the CLA correctly.

richkadel commented on src/parser.rs:
```diff
@@ -1504,8 +1480,8 @@ expected_indicator:    {}
         #[test]
         fn test_line_comment(
             line_comment_prefix in r#"//"#,
```
`#` characters not removed here?

richkadel commented on src/parser.rs:
```diff
@@ -1526,7 +1502,7 @@ expected_indicator:    {}
         #[test]
         fn test_empty_line_comment(
             line_comment_prefix in r#"//"#,
```
`#` characters not removed?

richkadel commented on src/parser.rs:
```diff
@@ -1547,7 +1523,7 @@ expected_indicator:    {}
         #[test]
         fn test_block_comment(
             block_comment_content in r#"([^*]|([*][^*/]))*"#,
```
`#` characters? I'm not sure why they are removed in many places, but not everywhere that doesn't have embedded doublequotes. (I prefer the consistency of always having them, but if they are going to be removed they should be removed consistently.)

richkadel commented on src/parser.rs:
```diff
@@ -1687,9 +1663,9 @@ expected_indicator:    {}
         #[test]
         fn test_single_quoted_string(
             single_quote in r#"'"#,
```
`#`

richkadel commented on src/parser.rs:
```diff
@@ -1739,7 +1715,7 @@ expected_indicator:    {}
                     r#"(null|true|false)|([-+]?(NaN|Infinity|(0[xX][0-9a-fA-F]+)"#,
                     r#"|((0|([1-9][0-9]*))?\.[0-9]+[eE][+-]?[0-9]+)|((0|([1-9][0-9]*))?\.[0-9]+)|((0|([1-9][0-9]*))\.?)))"#
```
`#`s

richkadel commented on src/parser.rs:
```diff
@@ -1774,7 +1750,7 @@ expected_indicator:    {}
         #[test]
         fn test_comma(
             comma in r#","#,
```
`#`

richkadel commented on src/parser.rs:
```diff
@@ -1789,10 +1765,10 @@ expected_indicator:    {}
         #[test]
         fn test_colon(
             colon in r#":"#,
```
`#`

richkadel reviewed (changes requested):
Thanks for the improvements to json5format! I added a few comments. Mostly nits plus an accidentally duplicated line.

I'm glad to see many of the simplifications. There are some changes that I'm not a huge fan of, but willing to accept clippy's suggestions. For what it's worth, I debated the:

`some_collection.len() > 0` vs. `!some_collection.is_empty()`

and thought the `!` is so easily overlooked, and so distant from `is_empty()` that I decided to use `len() > 0` for better readability.

But I'm willing to accept that change.

I'm less happy with removing the `#` from the raw strings, but again willing to accept them if the change is applied consistently.

If clippy is really warning about this, then fine, except I don't know why it's inconsistent. It left the `#` on many of the one-line raw-strings.

And similarly, replacying `r##"` (etc.) with `r#"` means I have to scan the multiline strings to see if there's a `"#` embedded somewhere, to make sure a single `#` is OK. I thought it was better to just keep them consistent `#` for single line raw strings and `##` for multi-line, without worrying about the string content (or almost never anyway). 

I'd be happy if you reverted the raw string `#"` changes, but if clippy is warning about these then I'm willing to accept it; HOWEVER, it should be consistently changed. I commented on several that I found where it was not.

Thanks so much though, and once you address the comments one way or the other, I can  approve.

bschwind commented:
@richkadel I felt similarly when going through the changes as I also prefer consistency on the `#` strings, and understand that the negation `!` character can be easily overlooked. We can opt out of those clippy rules with `allow` statements for clippy.

I'll play around with those and come up with something that matches the style you're going for.

bschwind committed:
Allow clippy lint for length comparisons to zero

bschwind commented:
@richkadel I have reverted the `#` changes in strings. My apologies, I think I may have been using a nightly toolchain which suggested to format those strings.

I also reverted the `!container.is_empty()` changes back to `container.len() > 0`. Would you also prefer to keep `container.len() == 0` calls instead of `container.is_empty()` ?

I can also rebase and squash commits as needed.

richkadel reviewed (approved):
LGTM but please squash, thanks!

bschwind commented:
I've rebased and organized commits into logical steps for this PR - I'm happy to squash to one commit if you prefer, but that can also be achieved by going to https://github.com/google/json5format/settings and allowing squash merging. That way the original commits in this PR will be preserved, while the history tree in `master` will only have one commit and stay linear. Your choice! :)

richkadel commented:
> going to https://github.com/google/json5format/settings and allowing squash merging

@erickt - It looks like I'm no longer an admin/maintainer since leaving Google. Can you check if this setting is enabled?

You (or another Google maintainer) may also need to approve the workflows. (I'm not sure what that's about, but the warning above says it requires a maintainer.)

And if you don't mind, I think you may need to do the merge and version update, and post to crates.io. If the crates.io requires permissions you don't have (assuming I still do), let me know.

Thanks!
